test/e2e/storage/testsuites/multiattach.go

```diff
@@ -0,0 +1,323 @@
+
+		// Test read/write work well in pod1 after pod2 is deleted
+		By("Rechecking if the volume in 1st pod exists as expected volume mode")
+		utils.CheckVolumeModeOfPath(pod1, l.resource1.pattern.VolMode, "/mnt/volume1")
+
+		By("Rechecking if read/write to the volume in 1st pod works properly")
+		utils.CheckReadWriteToPath(pod1, l.resource1.pattern.VolMode, "/mnt/volume1")
+	}
+
+	// This tests below configuration:
+	//          [pod1]
+	//          /    \      <- same volume mode
+	//   [volume1]  [volume2]
+	It("should attach two volumes with the same volume mode to the same pod", func() {
+		// Currently, multiple volumes are not generally available for pre-provisoined volume,
+		// because containerized storage servers, such as iSCSI and rbd, are just returning
+		// a static volume inside container, not actually creating a new volume per request.
+		if pattern.VolType == testpatterns.PreprovisionedPV {
+			framework.Skipf("This test doesn't work with pre-provisioned volume -- skipping")
+		}
+
+		init()
+		defer cleanup()
+
+		l.resource2.pattern.VolMode = l.resource1.pattern.VolMode
+		testAttachTwoVolumesToOnePod(l)
+	})
+
+	// This tests below configuration:
+	//          [pod1]
+	//          /    \      <- different volume mode (only <block, filesystem> pattern is tested)
+	//   [volume1]  [volume2]
+	It("should attach two volumes with different volume mode to the same pod", func() {
+		if pattern.VolMode == v1.PersistentVolumeFilesystem {
+			framework.Skipf("Filesystem volume case should be covered by block volume case -- skipping")
+		}
+
+		// Currently, multiple volumes are not generally available for pre-provisoined volume,
+		// because containerized storage servers, such as iSCSI and rbd, are just returning
+		// a static volume inside container, not actually creating a new volume per request.
+		if pattern.VolType == testpatterns.PreprovisionedPV {
+			framework.Skipf("This test doesn't work with pre-provisioned volume -- skipping")
+		}
+
+		init()
+		defer cleanup()
+
+		// Only Block volume test for 1st volume will reach here,
+		// so we should set filesystem for 2nd volume.
+		l.resource2.pattern.VolMode = v1.PersistentVolumeFilesystem
+		testAttachTwoVolumesToOnePod(l)
+	})
+
+	// This tests below configuration:
+	// [pod1] [pod2]
+	// [   node1   ]
+	//   \      /     <- same volume mode
+	//   [volume1]
+	It("should attach the same volume with the same volume mode to the different pod on the same node", func() {
+		init()
+		defer cleanup()
+
+		l.resource2.pattern.VolMode = l.resource1.pattern.VolMode
+		l.resource2.pvc = l.resource1.pvc
+		testAttachOneVolumeToTwoPods(l, true /* sameNode */)
+	})
+
+	// This tests below configuration:
+	//        [pod1] [pod2]
+	// [   node1   ] [   node2   ]
+	//         \      /     <- same volume mode
+	//         [volume1]
+	It("should attach the same volume with the same volume mode to the different pod on different node", func() {
```

Comment:
This test case is going to fail for RWO volumes. This test should be gated by a RWX volume capability.